### PR TITLE
ZNTA-1735-editor-glossary

### PR DESCRIPTION
**Pull request by kathryngo**
JIRA issue URL: https://zanata.atlassian.net/browse/ZNTA-1735

Add Glossary tab to Editor - and button next to TM Suggestions one with glossary count

**Note: The Activity Tab has been disabled before merging this into master because the content does not exist yet. ~~Or we need to add the Activity tab before merge~~** 

## TO DO: ##
~~- merge in master after [PR 211](https://github.com/zanata/zanata-platform/pull/211) verified and merged (this branch has been created from the above branch which are on QA so any changes will need to be added to this PR branch). - @kathryngo~~
~~- add button for glossary count next to TM suggestions - @kathryngo~~
- backend stuff - @davidmason 
~~- remove search dropdown - @kathryngo~~
~~- glossary entry modal - @kathryngo~~
~~- glossary table vertical scroll - @kathryngo~~
~~- enter a search term - add icon - @kathryngo~~
~- handle long glossary terms with ellipses - @kathryngo~~

Screenshot _3/3/17 12:25_

gloss

## Reviewer tasks

The following is based on the
[Zanata Development Guidelines](https://github.com/zanata/zanata-platform/wiki/Development-Guidelines).
As a reviewer, you should check the guidelines regularly for updates, and just
to refresh your memory.

The reviewer can use this list for reference to make sure
they have considered all the important points.

Use the checkboxes or not, whatever works best for you.

 - Code quality
   - [ ] Code passes style check (checkstyle/eslint)
   - [ ] Code is clear and concise
   - [ ] UI code is internationalised
   - [ ] Code has adequate comments where appropriate
   - Code is in an appropriate place
     - [ ] Classes are in appropriate packages
     - [ ] Code fits with class's responsibilities (SRP)
   - [ ] Configuration files are documented enough
   - If code is removed
     - [ ] No remaining code references the removed code
       - [ ] No orphaned rewrite rules
       - [ ] No orphaned config settings
     - [ ] No remaining comments refer to the removed code
     - [ ] Unused imports and libraries are removed
 - Testing
   - [ ] New code has tests
   - [ ] Changed code has tests
   - [ ] All tests pass
   - [ ] Test coverage stable or increased
 - Documentation
   - [ ] New features are documented
   - [ ] Docs removed for deleted features
   - [ ] Docs updated for all changed features
   - [ ] Developer/sysadmin docs updated if appropriate
 - If there are new dependencies
   - [ ] Does each new dependency pass all the
         [Considerations for new dependencies](https://github.com/zanata/zanata-platform/wiki/Development-Guidelines#new-technologies-and-dependencies)?

----
*This template can be updated in .github/PULL_REQUEST_TEMPLATE.md*

---
This change is [Reviewable](https://reviewable.io/reviews/zanata/zanata-platform/216)

### CODE REVIEW

**Comment by kathryngo** [comment]
@davidmason frontend stuff is finished on this branch - all those issues we discussed have been resolved.

**Comment by kathryngo** [comment]
@davidmason, I added some css to handle ellipses. If we want to show the ellipses in the middle of a long string then we will need to write some code to set a length of the string before and after the ellipses.

For the meantime I have a solution that shows long strings in full, and long words with ellipses.

Handling long strings:

- One long word - ellipses at end

 Thisisalong...

- Long string - ellipses at end, show all on hover

This is a lo...

On hover:

This is a long word
in the glossary

**Comment by kathryngo** [comment]
Added the Glossary Details Modal in storybook. Uses react-bootstrap components while still using the layout from the zanata-ui modals (ie. no footer, 'x' to close)

gloss-deets-modal

**Comment by seanf** [comment]
@zanata-jenkins test this please

**Comment by kathryngo** [comment]
@davidmason, I have made all those frontend changes as requested. Please get in touch asap if I have missed anything.

**Comment by kathryngo** [comment]
Scrollbar for table included
gloss-details-mod

**Comment by kathryngo** [comment]
@davidmason I have added all the components that have their own directory (editor/components/COMPONENT NAME/storyjs), and are used in the editor, to storybook. I just had to fix a few minor issues, it was mostly done. Thank you for putting the Glossay Term component in there.

**Comment by kathryngo** [comment]
@zanata-jenkins retest this please

**Comment by kathryngo** [comment]
@zanata-jenkins  retest this please

**Comment by kathryngo** [comment]
it's looking good 👍

**Comment by kathryngo** [comment]
@zanata-jenkins retest this please

**Comment by kathryngo** [comment]
@zanata-jenkins retest this please

File: server/zanata-frontend/src/frontend/app/components/RBButton/story.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import React from 'react'
+import { storiesOf, action } from '@kadira/storybook'
+import { Button } from 'react-bootstrap'
+
+storiesOf('Button', module)
```
Comment (kathryngo) [inline server/zanata-frontend/src/frontend/app/components/RBButton/story.js]: @davidmason I added this file when I was trying to create the combined editor/frontend storybook. 

You can go ahead and remove this (/RBButton).

File: server/zanata-frontend/src/frontend/app/components/stories.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+/**
+ * Collection of component stories for React Storybook.
+ *
+ * See .storybook for main config
+ */
+
+require('./RBButton/story.js')
```
Comment (kathryngo) [inline server/zanata-frontend/src/frontend/app/components/stories.js]: @davidmason and remove here too

**Comment by kathryngo** [comment]
Looks good @davidmason 👍

**Comment by aeng** [comment]
Reviewed 25 of 55 files at r1, 1 of 2 files at r2.
Review status: 25 of 52 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/GlossaryResource.java, line 210 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5Yk9GzBUrFnhhDjsI:-Kh5Yk9GzBUrFnhhDjsJ:bd8006c) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/api/zanata-common-api/src/main/java/org/zanata/rest/service/GlossaryResource.java#L210)):*
> ```Java
>     @Path("/details/{locale}")
>     @Produces({ MediaTypes.APPLICATION_ZANATA_GLOSSARY_JSON,
>             MediaType.APPLICATION_JSON })
> ```

Any typeHint for this?

---

*[server/zanata-frontend/src/frontend/package.json, line 111 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5ae_1k3s7Xh1JGGjg:-Kh5ae_2jZAZBfDS49-f:b-57w9xv) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-frontend/src/frontend/package.json#L111)):*
> ```JSON
>     "react-list": "0.7.19",
>     "react-loading": "0.0.9",
>     "react-lorem-component": "^0.11.0",
> ```

exact version?

---

*[server/zanata-frontend/src/frontend/app/editor/actions/glossary.js, line 39 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5beZVZjt5DZOJDatc:-Kh5beZVZjt5DZOJDatd:b-1z4mu) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-frontend/src/frontend/app/editor/actions/glossary.js#L39)):*
> ```JavaScript
>     waitForPhraseDetail(getState, phraseId, (phrase) => {
>       dispatch(glossarySearchTextEntered(phrase.sources.join(' ')))
>     }, 20, () => {
> ```

what's this 20?

---

*[server/zanata-frontend/src/frontend/app/editor/actions/glossary.js, line 92 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5bjR4wFg9kQRUM-hL:-Kh5bjR4wFg9kQRUM-hM:b15lnj2) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-frontend/src/frontend/app/editor/actions/glossary.js#L92)):*
> ```JavaScript
> 
>     const glossaryUrl =
>       `${baseRestUrl}/glossary/search?srcLocale=${srcLocale}&transLocale=${transLocale}&project=${projectSlug}&searchText=${encodeURIComponent(searchText)}&maxResults=15` // eslint-disable-line max-len
> ```

15 results by default? Can we make the 15 a const?

---

*[server/zanata-frontend/src/frontend/app/editor/actions/suggestions.js, line 256 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5hlDDeKcE6EVsjNdl:-Kh5hlDDeKcE6EVsjNdm:bfummsa) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-frontend/src/frontend/app/editor/actions/suggestions.js#L256)):*
> ```JavaScript
>     waitForPhraseDetail(getState, phraseId, (phrase) => {
>       dispatch(findPhraseSuggestions(phrase))
>     }, 20, () => {
> ```

can have the 20 as const?

---

*[server/zanata-frontend/src/frontend/app/editor/components/GlossarySearchInput/index.js, line 35 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5llptdZ9P27oaj4gl:-Kh5llptdZ9P27oaj4gm:b-iqkeht) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-frontend/src/frontend/app/editor/components/GlossarySearchInput/index.js#L35)):*
> ```JavaScript
>     text: PropTypes.string.isRequired,
>     onTextChange: PropTypes.func.isRequired
>     // loading: PropTypes.bool.isRequired,
> ```

Still need those?

---

*[server/zanata-frontend/src/frontend/app/editor/components/GlossaryTermModal/component.js, line 1 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5tXt5Y7eP0hoql_MC:-Kh5tXt5Y7eP0hoql_MD:bav37au) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-frontend/src/frontend/app/editor/components/GlossaryTermModal/component.js#L1)):*
> ```JavaScript
> import React, { PropTypes } from 'react'
> ```

component within component... interesting

---

*[server/zanata-frontend/src/frontend/app/editor/components/GlossaryTermModal/index.css, line 3 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5tcLlyn4mIjLblu-V:-Kh5tcLlyn4mIjLblu-W:b-piql3h) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-frontend/src/frontend/app/editor/components/GlossaryTermModal/index.css#L3)):*
> ```CSS
> 
> /* Shorten with an ellipsis when overflowing */
> .ellipsis {
> ```

dont we have something similar in the general css? or we should move this class to the parent css.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/216)*

**Comment by kathryngo** [comment]
*[server/zanata-frontend/src/frontend/app/editor/components/GlossaryTermModal/index.css, line 3 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5tcLlyn4mIjLblu-V:-Kh5vTCafq3PfBW7Q-oi:b-4vybat) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-frontend/src/frontend/app/editor/components/GlossaryTermModal/index.css#L3)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

dont we have something similar in the general css? or we should move this class to the parent css.
</blockquote></details>

In the frontend css yes, which editor doesn't use since it overrides a lot of exiting styles. Had to cherrypick css for the editor design.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/216)*

**Comment by kathryngo** [comment]
*[server/zanata-frontend/src/frontend/app/editor/components/GlossaryTermModal/index.css, line 3 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5tcLlyn4mIjLblu-V:-Kh5wA--3YKjuFhalBkD:b-hhbp72) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-frontend/src/frontend/app/editor/components/GlossaryTermModal/index.css#L3)):*
<details><summary><i>Previously, kathryngo wrote…</i></summary><blockquote>

In the frontend css yes, which editor doesn't use since it overrides a lot of exiting styles. Had to cherrypick css for the editor design.
</blockquote></details>

I will move the ellipsis class to the root/index.css file though (parent css)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/216)*

**Comment by aeng** [comment]
Reviewed 16 of 55 files at r1, 1 of 2 files at r2, 1 of 1 files at r3.
Review status: 41 of 52 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[server/zanata-frontend/src/frontend/app/editor/containers/GlossaryTab.js, line 119 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5umvNP-JGaS4tLOrD:-Kh5umvNP-JGaS4tLOrE:b-yblobv) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-frontend/src/frontend/app/editor/containers/GlossaryTab.js#L119)):*
> ```JavaScript
> 
> function mapStateToProps ({ glossary }) {
>   // FIXME move this to storybook for this component
> ```

Should clean up this section? or wait til storybook is setup?

---

*[server/zanata-frontend/src/frontend/app/editor/containers/SidebarContent.js, line 105 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5v-gX1vG8dEcGLQgt:-Kh5v-gX1vG8dEcGLQgu:b7k6v79) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-frontend/src/frontend/app/editor/containers/SidebarContent.js#L105)):*
> ```JavaScript
>     const { glossaryCount } = this.props
>     const glossaryCountDisplay = glossaryCount > 0
>       // TODO kgough display as a badge instead of text in parens
> ```

Something missed out to implement?

---

*[server/zanata-war/src/main/java/org/zanata/rest/service/GlossaryService.java, line 189 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5w6wrk6dYf1rbByv3:-Kh5w6wrk6dYf1rbByv4:b-2h42ki) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-war/src/main/java/org/zanata/rest/service/GlossaryService.java#L189)):*
> ```Java
>                     .build());
>         }
>         if (maxResults < 1 || maxResults > 1000) {
> ```

1000 should be a static final int

---

*[server/zanata-war/src/main/java/org/zanata/service/impl/GlossarySearchServiceImpl.java, line 83 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5x4bdZWbd6luOKWDg:-Kh5x4bezY9YQrMitoMZ:b-q4q47o) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-war/src/main/java/org/zanata/service/impl/GlossarySearchServiceImpl.java#L83)):*
> ```Java
>                     Maps.newLinkedHashMap();
> 
>             if (projectSlug != null) {
> ```

does CheckForNull prevent this from nullable?

---

*[server/zanata-war/src/main/java/org/zanata/webtrans/server/rpc/GetGlossaryHandler.java, line 67 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5xcO_s555sQrk8fn6:-Kh5xcO_s555sQrk8fn7:b8cwp01) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-war/src/main/java/org/zanata/webtrans/server/rpc/GetGlossaryHandler.java#L67)):*
> ```Java
>         ArrayList<GlossaryResultItem> results;
>         try {
>             results = glossarySearchService.searchGlossary(
> ```

Nicely done on the refactoring.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/216)*

**Comment by kathryngo** [comment]
*[server/zanata-frontend/src/frontend/app/editor/containers/SidebarContent.js, line 105 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5v-gX1vG8dEcGLQgt:-Kh5yagE56RJIEXNhnyz:b-bessbb) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-frontend/src/frontend/app/editor/containers/SidebarContent.js#L105)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

Something missed out to implement?
</blockquote></details>

Badge/label to show glossary count next to Glossary title. Simple addition that can be made at a later date.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/216)*

**Comment by kathryngo** [comment]
*[server/zanata-frontend/src/frontend/app/editor/components/GlossaryTermModal/component.js, line 1 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5tXt5Y7eP0hoql_MC:-Kh5zLaMCp1dgG7TrWCc:bglvyua) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-frontend/src/frontend/app/editor/components/GlossaryTermModal/component.js#L1)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

component within component... interesting
</blockquote></details>

By that do you mean the filename should be index.js? This isn't my work but I am curious.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/216)*

**Comment by kathryngo** [comment]
*[server/zanata-frontend/src/frontend/app/editor/containers/GlossaryTab.js, line 119 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5umvNP-JGaS4tLOrD:-KhMBcQMcscLPFthIBp_:b58urb9) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-frontend/src/frontend/app/editor/containers/GlossaryTab.js#L119)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

Should clean up this section? or wait til storybook is setup?
</blockquote></details>

I'd wait. 
Btw, have answered all I feel confident answering in this review. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/216)*

**Comment by kathryngo** [comment]
@davidmason looks good besides one issue. The modal is not working.
When I click the info icon, the modal does not pop up and the glossary tab layout changes (the sidebar jumps to appear under table):

glossbug

**Comment by kathryngo** [comment]
:lgtm: 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-:-KhPWBBJ9M2X96rCcUv4:bg7igkv)*

**Comment by davidmason** [comment]
Review status: 40 of 55 files reviewed at latest revision, 13 unresolved discussions.

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/GlossaryResource.java, line 210 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5Yk9GzBUrFnhhDjsI:-KhPYnbyYBjG5MPc4MwD:b-wjtdo7) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/api/zanata-common-api/src/main/java/org/zanata/rest/service/GlossaryResource.java#L210)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

Any typeHint for this?
</blockquote></details>

This is a tricky one because the payload is one of the GWT model classes. To add a `@TypeHint` here the options are:

- add a circular dependency from zanata-common-api to zanata-war (seems like a pretty bad idea)
- move the payload type into zanata-common-api, but this also means it would add a dependency on GWT for that module since the payload type needs the gwt `isSerializable` type
- define a new type without the gwt isSerializable and write a converter to switch between them

But the easiest option is to not use a TypeHint yet, and just wait until we remove GWT before we move the resource class and add the TypeHint.

What do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/216)*

**Comment by aeng** [comment]
Reviewed 2 of 55 files at r1, 2 of 2 files at r4, 1 of 3 files at r5, 1 of 2 files at r6, 1 of 2 files at r7.
Review status: 47 of 55 files reviewed at latest revision, 10 unresolved discussions.

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/GlossaryResource.java, line 210 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5Yk9GzBUrFnhhDjsI:-KhP_sdoONwpapYWWobs:b-uwf8tz) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/api/zanata-common-api/src/main/java/org/zanata/rest/service/GlossaryResource.java#L210)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

This is a tricky one because the payload is one of the GWT model classes. To add a `@TypeHint` here the options are:

- add a circular dependency from zanata-common-api to zanata-war (seems like a pretty bad idea)
- move the payload type into zanata-common-api, but this also means it would add a dependency on GWT for that module since the payload type needs the gwt `isSerializable` type
- define a new type without the gwt isSerializable and write a converter to switch between them

But the easiest option is to not use a TypeHint yet, and just wait until we remove GWT before we move the resource class and add the TypeHint.

What do you think?
</blockquote></details>

Alright. Perhaps TODO to include typehint in the future?

---

*[server/zanata-war/src/test/java/org/zanata/webtrans/server/rpc/GetGlossaryDetailsHandlerTest.java, line 106 at r7](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-KhPaqumtIZRgF0KsPJ4:-KhPaqun2595eqdPLDBF:bt6nz0g) ([raw file](https://github.com/zanata/zanata-platform/blob/b7d884ecb892427fd21519783a5bfe4eee7e9f96/server/zanata-war/src/test/java/org/zanata/webtrans/server/rpc/GetGlossaryDetailsHandlerTest.java#L106)):*
> ```Java
> //    @Test(expected = ActionException.class)
> //    @InRequestScope
> //    public void testExecuteWithInvalidLocale() throws Exception {
> ```

Any reason for commenting out this test?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/216)*

**Comment by davidmason** [comment]
Review status: 47 of 55 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[api/zanata-common-api/src/main/java/org/zanata/rest/service/GlossaryResource.java, line 210 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5Yk9GzBUrFnhhDjsI:-KhPf_Z1oMAf9E7AF2xS:b-ovwaw7) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/api/zanata-common-api/src/main/java/org/zanata/rest/service/GlossaryResource.java#L210)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

Alright. Perhaps TODO to include typehint in the future?
</blockquote></details>

Good idea.

---

*[server/zanata-frontend/src/frontend/app/editor/actions/glossary.js, line 39 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5beZVZjt5DZOJDatc:-KhPbyYv2Zjv19qWLOgh:blta4d8) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-frontend/src/frontend/app/editor/actions/glossary.js#L39)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

what's this 20?
</blockquote></details>

How many times to try polling for the phrase detail before it gives up and logs an error. Extracting a const for it.

---

*[server/zanata-frontend/src/frontend/app/editor/components/GlossarySearchInput/index.js, line 35 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5llptdZ9P27oaj4gl:-KhPdnzQePzWHw5era5c:b-xd4yip) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-frontend/src/frontend/app/editor/components/GlossarySearchInput/index.js#L35)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

Still need those?
</blockquote></details>

Nope, removing :)

---

*[server/zanata-war/src/main/java/org/zanata/service/impl/GlossarySearchServiceImpl.java, line 83 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5x4bdZWbd6luOKWDg:-KhPeoLar28ALMeSSWsY:bdsegcj) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-war/src/main/java/org/zanata/service/impl/GlossarySearchServiceImpl.java#L83)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

does CheckForNull prevent this from nullable?
</blockquote></details>

`@CheckForNull` means it may be null and we need to check it. According to Sean it is the best one to use at the moment when we expect something to possibly be null (and `@Nonnull` is the best when we should never have a null value passed in there).

---

*[server/zanata-war/src/main/java/org/zanata/webtrans/server/rpc/GetGlossaryHandler.java, line 67 at r3](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-Kh5xcO_s555sQrk8fn6:-KhPfCOeYiFKW-LMPTFH:bvtk9y9) ([raw file](https://github.com/zanata/zanata-platform/blob/28f9c3f5d9dcda443e8f67f2150cc51546d4d157/server/zanata-war/src/main/java/org/zanata/webtrans/server/rpc/GetGlossaryHandler.java#L67)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

Nicely done on the refactoring.
</blockquote></details>

Thanks :)

---

*[server/zanata-war/src/test/java/org/zanata/webtrans/server/rpc/GetGlossaryDetailsHandlerTest.java, line 106 at r7](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-KhPaqumtIZRgF0KsPJ4:-KhPfH92_AGPMBSIP-Ez:b-ucie1q) ([raw file](https://github.com/zanata/zanata-platform/blob/b7d884ecb892427fd21519783a5bfe4eee7e9f96/server/zanata-war/src/test/java/org/zanata/webtrans/server/rpc/GetGlossaryDetailsHandlerTest.java#L106)):*
<details><summary><i>Previously, aeng (Alex Eng) wrote…</i></summary><blockquote>

Any reason for commenting out this test?
</blockquote></details>

It doesn't check that any more, so it should now just give empty results if there's nothing for that locale. I could update the test to expect nothing back, or just remove it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/216)*

**Comment by davidmason** [comment]
Review status: 47 of 55 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[server/zanata-war/src/test/java/org/zanata/webtrans/server/rpc/GetGlossaryDetailsHandlerTest.java, line 106 at r7](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-KhPaqumtIZRgF0KsPJ4:-KhPviZ0gXMuaSDFX753:boxlqbl) ([raw file](https://github.com/zanata/zanata-platform/blob/b7d884ecb892427fd21519783a5bfe4eee7e9f96/server/zanata-war/src/test/java/org/zanata/webtrans/server/rpc/GetGlossaryDetailsHandlerTest.java#L106)):*
<details><summary><i>Previously, davidmason (David Mason) wrote…</i></summary><blockquote>

It doesn't check that any more, so it should now just give empty results if there's nothing for that locale. I could update the test to expect nothing back, or just remove it.
</blockquote></details>

I had a closer look and it isn't even calling the method that would throw an exception any more, so I think it is best just to remove the test as it no longer makes sense.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/216)*

**Comment by aeng** [comment]
:lgtm:

---

Reviewed 1 of 2 files at r7, 6 of 7 files at r8.
Review status: 54 of 55 files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-:-KhPxzCa1vABNKJRmNBJ:bnfp4nl)*

**Comment by davidmason** [comment]
Reviewed 1 of 7 files at r8.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/216)*

**Comment by spathare21** [comment]
HI ,

I have tested this functionality and observer below things, other than that all works fine.
- glossary source , target  and info icon are not aligned properly. 

screenshot from 2017-04-11 09-06-50


- why there is two icons (square and i i.e glossary and info icon) to display same thing i.e glossary 


screenshot from 2017-04-11 09-07-47


- once i click on copy the target string have to be copy in target translation but nothing happed on copy click.

**Comment by davidmason** [comment]
@spathare21 which browser are you testing this in? Do you see the same layout problem with other browsers?

Does your test data have any target content for the language you have selected? I do not see any target text, and that may be causing the layout issue and it would explain why no text is inserted if you click Copy.

> * why there is two icons (square and i i.e glossary and info icon) to display same thing i.e glossary

One icon is for the details panel, the other is for the glossary. There will soon be more tabs in addition to glossary so the separate icon will make more sense then.

**Comment by kathryngo** [comment]
@spathare21, fixing the table layout now.

**Comment by spathare21** [comment]
Verified , now all the changes are working fine.

**Comment by kathryngo** [comment]
@aeng minor css changes have been made, as well as changes to other files (upgrading yarn it looks like). Would you mind reviewing these changes please?

**Comment by davidmason** [comment]
@zanata-jenkins retest this please

**Comment by aeng** [comment]
Reviewed 4 of 4 files at r10, 1 of 2 files at r12, 3 of 3 files at r13.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/216)*

**Comment by aeng** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/zanata/zanata-platform/216#-:-KhU5YWn3IFdThzPmrCD:bnfp4nl)*

**Comment by davidmason** [comment]
@zanata-jenkins retest this please

**Comment by davidmason** [comment]
@zanata-jenkins retest this please

**Pull request merged**